### PR TITLE
SP-266 - Get Transaction

### PR DESCRIPTION
**Pull request by RaymondK-mx51**
Thanks for reviewing this.

### CODE REVIEW

**Comment by RaymondK-mx51** [comment]
I tested some edge cases as well as against the Recovery test cases in the cert. 

Thanks Pat for reviewing this. Let me know what you find.

File: SPIClient/Spi.cs [hunk 1]
```diff
@@ -1456,6 +1485,114 @@ private void _handleErrorEvent(Message m)
...
+                    }
+                    else if (gtResponse.PosRefIdMissing())
+                    {
+                        Log.Information($"Get transaction did not match, PosRefId is invalid.");
+                        txState.UnknownCompleted("Failed to recover Transaction Status. Check EFTPOS. ");
+                    }
+                }
+                else
+                {
+                    if (txState.AttemptingToRecover == true)
```
Comment (ellulpatrick) [inline SPIClient/Spi.cs:1565]: I'm not quite sure why we need this distinction on `AttemptingToRecover`. I can see that both paths have basically the same code. When it's PosCrashRecovery we also still have to consider unknown. what was the main reason why you needed this flag? We didn;t have the concept before, right?

File: SPIClient/Purchase.cs [hunk 2]
```diff
@@ -281,6 +281,130 @@ public string GetResponseValueWithAttribute(string attribute)
...
+        }
+
+        public bool WasTransactionInProgressError()
+        {
+            return _m.GetError().StartsWith("TRANSACTION_IN_PROGRESS");
+        }
+        
+        public bool IsStillInProgress(string posRefId)
+        {
+            return (WasOperationInProgressError() || WasTransactionInProgressError()) && (GetPosRefId().Equals(posRefId) || GetPosRefId() == null);
```
Comment (ellulpatrick) [inline SPIClient/Purchase.cs:384]: not quite understand this... i thought that GT will only return TransactionInProgress if the one that was asked for is in progress.... so there is no need to match posrefid?

File: SPIClient/Spi.cs [hunk 3]
```diff
@@ -1456,6 +1485,114 @@ private void _handleErrorEvent(Message m)
...
+                }
+
+                Log.Information($"Got Transaction response.");
+                txState.GotGtResponse(); 
+                var gtResponse = new GetTransactionResponse(m);
+
+                // check if GT is returned successfully
+                if (!gtResponse.WasRetrievedSuccessfully())
+                {
+                    if (gtResponse.IsStillInProgress(txState.PosRefId))
```
Comment (ellulpatrick) [inline SPIClient/Spi.cs:1522]: do we need to check against posrefid here? 
my understanding is that GT, if there is anything in progress, it either:
```
if the one asked for is in progress, 
     then it returns in_progress error
else if it find it in it store all done and dusted,
     it returns it from the store
else 
    it returns a not_found error
```

File: SPIClient/Spi.cs [hunk 4]
```diff
@@ -1456,6 +1485,114 @@ private void _handleErrorEvent(Message m)
...
+                        else
+                        {
+                            Log.Information($"Operation still in progress... stay waiting.");
+                            return;
+                        }
+                    }
+                    else if (gtResponse.PosRefIdNotFound()) 
+                    {
+                        Log.Information($"Get transaction did not match.");
+                        txState.UnknownCompleted("Failed to recover Transaction Status. Check EFTPOS. ");
```
Comment (ellulpatrick) [inline SPIClient/Spi.cs:1550]: this is now where we can remove an Unknown Case. It should instead be a Completed, with Failed state. as we can be sure this txn did not happen, and POS can retry,

File: SPIClient/Spi.cs [hunk 5]
```diff
@@ -1456,6 +1485,114 @@ private void _handleErrorEvent(Message m)
...
+                    }
+                    else if (gtResponse.PosRefIdInvalid())
+                    {
+                        Log.Information($"Get transaction did not match, PosRefId is invalid.");
+                        txState.UnknownCompleted("Failed to recover Transaction Status. Check EFTPOS. ");
```
Comment (ellulpatrick) [inline SPIClient/Spi.cs:1555]: this can also be considered as a Failed Completed, not as an Unknown

File: SPIClient/Spi.cs [hunk 6]
```diff
@@ -1456,6 +1485,114 @@ private void _handleErrorEvent(Message m)
...
+                    }
+                    else if (gtResponse.PosRefIdMissing())
+                    {
+                        Log.Information($"Get transaction did not match, PosRefId is invalid.");
+                        txState.UnknownCompleted("Failed to recover Transaction Status. Check EFTPOS. ");
```
Comment (ellulpatrick) [inline SPIClient/Spi.cs:1560]: this can also be considered as a Failed Completed, not as an Unknown

File: SPIClient/Spi.cs [hunk 7]
```diff
@@ -1456,6 +1485,114 @@ private void _handleErrorEvent(Message m)
...
+
+                        if (tx != null)
+                        {
+                            Log.Information("$Retrieved transaction during recovery.");
+                            gtResponse.CopyMerchantReceiptToCustomerReceipt();
+                            txState.Completed(tx.GetSuccessState(), tx, "Transaction Ended.");
+                        }
+                        else
+                        {
+                            // unexpected response during gt
```
Comment (ellulpatrick) [inline SPIClient/Spi.cs:1586]: maybe add more path context to the comment and log. basically if we're here it means that for some strange reason, VAA told us "success" but the "tx" element was missing. right? i.e. indicates a vaa bug?

File: SPIClient/Spi.cs [hunk 8]
```diff
@@ -1513,14 +1650,6 @@ private void _handleGetLastTransactionResponse(Message m)
                             return;
                         }
                     }
-                    else if (gtlResponse.WasTimeOutOfSyncError())
```
Comment (ellulpatrick) [inline SPIClient/Spi.cs:1516]: _handleGetLastTransaction should now be totally cleaned up from all recovery logic. it should simply check that it is being called when an explicit GLT was called from POS, and it should also fail very easily.

we have to test what happens when a public GLT call ends up itself being recovered by the timers using GT, but that shouldn't matter much.

Comment (ellulpatrick) [reply SPIClient/Spi.cs:1516]: also, it seems that `_callGetLastTransaction` is now dead code

File: SPIClient/Spi.cs [hunk 9]
```diff
@@ -1456,6 +1485,114 @@ private void _handleErrorEvent(Message m)
...
+                            CurrentTxFlowState.PhoneForAuthRequired(new PhoneForAuthRequired(txState.PosRefId, m.Id, "UNKNOWN", "UNKNOWN"), "Recovered mid Phone-For-Auth but don't have details. You may Cancel then Retry.");
+                        }
+                        else if (gtResponse.WasTransactionInProgressError())
+                        {
+                            Log.Information($"Transaction is currently in progress... stay waiting.");
+                            return;
+                        }
+                        else
+                        {
+                            Log.Information($"Operation still in progress... stay waiting.");
```
Comment (ellulpatrick) [inline SPIClient/Spi.cs:1543]: what is this path? what could trigger it?

File: SPIClient/Spi.cs [hunk 10]
```diff
@@ -1456,6 +1485,114 @@ private void _handleErrorEvent(Message m)
...
+                        }
+                        else if (gtResponse.IsWaitingForAuthCode() && !txState.AwaitingPhoneForAuth)
+                        {
+                            Log.Information($"Eftpos is waiting for us to send it auth code, but we were not aware of this. " +
+                                      $"We can only cancel the transaction at this stage as we don't have enough information to recover from this.");
+                            CurrentTxFlowState.PhoneForAuthRequired(new PhoneForAuthRequired(txState.PosRefId, m.Id, "UNKNOWN", "UNKNOWN"), "Recovered mid Phone-For-Auth but don't have details. You may Cancel then Retry.");
+                        }
+                        else if (gtResponse.WasTransactionInProgressError())
+                        {
+                            Log.Information($"Transaction is currently in progress... stay waiting.");
```
Comment (ellulpatrick) [inline SPIClient/Spi.cs:1538]: isn't this condition always true? because we enter this bock only if transaction is in progress in L1522?

File: SPIClient/Spi.cs [hunk 11]
```diff
@@ -1456,6 +1485,114 @@ private void _handleErrorEvent(Message m)
...
+                    else if (gtResponse.PosRefIdInvalid())
+                    {
+                        Log.Information($"Get transaction did not match, PosRefId is invalid.");
+                        txState.UnknownCompleted("Failed to recover Transaction Status. Check EFTPOS. ");
+                    }
+                    else if (gtResponse.PosRefIdMissing())
+                    {
+                        Log.Information($"Get transaction did not match, PosRefId is invalid.");
+                        txState.UnknownCompleted("Failed to recover Transaction Status. Check EFTPOS. ");
+                    }
```
Comment (ellulpatrick) [inline SPIClient/Spi.cs:1561]: we should probably have another catch all else here, which would mean that `!gtResponse.WasRetrievedSuccessfully()` but "none of the above" and try to reason what would bring us there....

File: SPIClient/SpiModels.cs [hunk 12]
```diff
@@ -234,11 +235,21 @@ public class TransactionFlowState : EventArgs
...
         public DateTime LastStateRequestTime { get; internal set; }
 
+        [Obsolete("Use Get Transaction instead, GtRequestId.")]
         /// <summary>
         /// The id of the last glt request message that was sent. used to match with the response.
         /// </summary>
         public string LastGltRequestId { get; internal set; }
 
+        /// <summary>
+        /// The id of the last glt request message that was sent. used to match with the response.
```
[the remark this answers is not in the document]
Comment (RaymondK-mx51) [reply SPIClient/SpiModels.cs]: Done

File: SPIClient/Spi.cs [hunk 13]
```diff
@@ -1085,7 +1114,7 @@ public Message.SuccessState GltMatch(GetLastTransactionResponse gltResponse, str
             return gltResponse.GetSuccessState();
         }
 
-        [Obsolete("Use GltMatch(GetLastTransactionResponse gltResponse, string posRefId, TransactionType expectedType)")]
+        [Obsolete("This should no longer be used as Get Last Transaction is replaced by Get Transaction")]
```
[the remark this answers is not in the document]
Comment (RaymondK-mx51) [reply SPIClient/Spi.cs:1117]: Brayden mentioned some PoS are using it, but not the exact number as you mentioned. They seem reluctant to make another breaking change. 

My concern obviously is breaking backwards compatibility. We have the patch coming up, so getting POS to rewire now/in the future is more work for them. I personally want to clean up and remove the code, so I will leave GLT for now and have a chat with them next week.

File: SPIClient/SpiModels.cs [hunk 14]
```diff
@@ -350,6 +366,29 @@ internal void GotGltResponse()
             AwaitingGltResponse = false;
         }
 
+        internal void CallingGt(string gtRequestId, bool attemptingRecovery)
```
[the remark this answers is not in the document]
Comment (RaymondK-mx51) [reply SPIClient/SpiModels.cs]: Agree, makes more sense when I am re-reading the code and I think I am over complicating it.

Initially I did this so that we can be explicit in the code & logs about about a explicit POS recovery. Especially the logs will be useful for troubleshooting. After re-reading, I think its enough to deduce explicit pos recovery from the first log.

File: SPIClient/SpiModels.cs [hunk 15]
```diff
@@ -234,11 +235,21 @@ public class TransactionFlowState : EventArgs
...
         public string LastGltRequestId { get; internal set; }
 
+        /// <summary>
+        /// The id of the last glt request message that was sent. used to match with the response.
+        /// </summary>
+        public string GtRequestId { get; internal set; }
+
+        /// <summary>
+        /// </summary>
+        public bool AttemptingToRecover { get; internal set; }
```
[the remark this answers is not in the document]
Comment (RaymondK-mx51) [reply SPIClient/SpiModels.cs]: Removed, see above comment

File: SPIClient/Spi.cs (hunk 1 above)
Comment (RaymondK-mx51) [reply SPIClient/Spi.cs]: Removed, see above comment

File: SPIClient/Purchase.cs (hunk 2 above)
Comment (RaymondK-mx51) [reply SPIClient/Purchase.cs]: Agree and done.

File: SPIClient/Spi.cs (hunk 3 above)
Comment (RaymondK-mx51) [reply SPIClient/Spi.cs]: Agree, fixed as part of above.

File: SPIClient/Spi.cs (hunk 4 above)
Comment (RaymondK-mx51) [reply SPIClient/Spi.cs]: Yes silly me :disappointed:

File: SPIClient/Spi.cs (hunk 5 above)
Comment (RaymondK-mx51) [reply SPIClient/Spi.cs]: Yes silly me :disappointed:

File: SPIClient/Spi.cs (hunk 6 above)
Comment (RaymondK-mx51) [reply SPIClient/Spi.cs]: Yes silly me :disappointed:

File: SPIClient/Spi.cs (hunk 7 above)
Comment (RaymondK-mx51) [reply SPIClient/Spi.cs]: Following comments was added:
// tx payload missing from get transaction protocol, could be a VAA issue

File: SPIClient/Spi.cs (hunk 8 above)
Comment (RaymondK-mx51) [reply SPIClient/Spi.cs:1516]: 1. Removed `_callGetLastTransaction`

2. I am leaving GLT as it is for now, I will follow up with Nick and Brayden next week to discuss.

File: SPIClient/Spi.cs (hunk 9 above)
Comment (RaymondK-mx51) [reply SPIClient/Spi.cs:1543]: If a GT call is made while the terminal is busy (ie. menu screen)?

File: SPIClient/Spi.cs (hunk 10 above)
Comment (RaymondK-mx51) [reply SPIClient/Spi.cs:1538]: Yes, it is true, I just want the right information to be shown in the log, so a transaction in progress vs operation in progress error, relates to the comment above. 

Happy to remove it if it clutters the code.

File: SPIClient/Spi.cs (hunk 11 above)
Comment (RaymondK-mx51) [reply SPIClient/Spi.cs:1567]: Agree and done.

**Comment by RaymondK-mx51** [comment]
Fixed all the comments except for GLT, will talk to the team about it next week and clean it up in one commit.

Sorry re. the unknown stuff!

File: SPIClient/Spi.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (ellulpatrick) [reply SPIClient/Spi.cs:1117]: It's not the GLT that I would like to remove... I'm ok to leave that there. It's the GltMatch because that always made wrong assumptions... and anyone that is using it means they're targeting a specific PosRefID. Which means it is easy enough to replace that GLT call with a GT.

[the remark this answers is not in the document]
Comment (ellulpatrick) [reply SPIClient/Spi.cs:1117]: Also, speaking of this, we need analytics on who's using what... we should start adding some...

File: SPIClient/Spi.cs (hunk 8 above)
Comment (ellulpatrick) [reply SPIClient/Spi.cs:1516]: I'm not saying that GLT should be removed. sorry didn;t explain myself well. I meant that _handleGLTResponse can be totally cleaned up from recovery logic. as it is not used for recovery anymore. (all that logic has moved to _handleGTResponse)

File: SPIClient/Spi.cs (hunk 10 above)
Comment (ellulpatrick) [reply SPIClient/Spi.cs:1538]: ok i see what's happening here... i think we can simplify it a little.... lat's talk in person....

**Comment by RaymondK-mx51** [comment]
> looking good.
> I've made further GLT cleanup recommendations on a new branch....
> And also realised that GLT itself cannot be recovered with a GT. so keeping it happy for now with a workaround in monitoring thread.

Yes, I just realised this. Happy path works for GLT, but cannot be recovered for anything else.
